common/src/main/java/io/druid/math/expr/Expr.java

```diff
@@ -258,320 +284,348 @@ public String toString()
   }
 }
 
-class BinMinusExpr extends BinaryOpExprBase
+abstract class BinaryNumericOpExprBase extends BinaryOpExprBase
 {
-
-  BinMinusExpr(String op, Expr left, Expr right)
+  public BinaryNumericOpExprBase(String op, Expr left, Expr right)
   {
     super(op, left, right);
   }
 
   @Override
-  public Number eval(ObjectBinding bindings)
+  public ExprEval eval(ObjectBinding bindings)
   {
-    Number leftVal = left.eval(bindings);
-    Number rightVal = right.eval(bindings);
-    if (isLong(leftVal, rightVal)) {
-      return leftVal.longValue() - rightVal.longValue();
-    } else {
-      return leftVal.doubleValue() - rightVal.doubleValue();
+    ExprEval leftVal = left.eval(bindings);
+    ExprEval rightVal = right.eval(bindings);
```

Comment:
as opposed to checking for type everywhere, can we use polymorphic deserialization instead? As in, we have different types of ExprEvals that are automatically instantiated based on Expr type?